### ISSUE TITLE
`Union[A, B, C]` gives different output to `Union[A, Union[B, C]]`

### ISSUE THREAD

**Issue by samuelcolvin** [microsoft/pyright#3983]
**Describe the bug**

See https://github.com/pydantic/pydantic-core/pull/248 and in particular https://github.com/pydantic/pydantic-core/pull/248/commits/ce0dce661d4d4cf313a607079b732d63a40f553a.

pyright is giving different results depending on whether I use

```py
PyrightPasses = Union[A, B, C]
# or
PyrightFailes = Union[A, Union[B, C]]
```

In my case `A` is a literal of strings, `B`, `C` and others are many `TypedDict`s.

**To Reproduce**

I've tried to create a minimal example, but I didn't see the same behaviour. I think the problem might related to how big/deep the types are?

**Expected behavior**

I think pyright should pass both before and after https://github.com/pydantic/pydantic-core/pull/248/commits/ce0dce661d4d4cf313a607079b732d63a40f553a.

**Screenshots or Code**

Best to look at the commit above.

**Comment by erictraut** [comment]
What is the error that it gives in the failure case?

A minimal example would be greatly appreciated. 

If that's not possible, could you provide more detailed instructions for how to repro the problem with the full pydantic code base?

**Comment by samuelcolvin** [comment]
I totally agree that a minimal example would be great, I tried to create one, but couldn't.

I forgot to add, that this is using `pyright==1.1.246`, I tried latest pright but:
1. The error remained
2. New errors were introduced, I wanted to fix this one before upgrading

The error can be seen [here](https://github.com/pydantic/pydantic-core/actions/runs/3122291934/jobs/5064137152), output is:

<details>

```
/home/runner/work/pydantic-core/pydantic-core/pydantic_core/__init__.py
  /home/runner/work/pydantic-core/pydantic-core/pydantic_core/__init__.py:1:6 - warning: Import "._pydantic_core" could not be resolved from source (reportMissingModuleSource)
/home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py
  /home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py:17:[26](https://github.com/pydantic/pydantic-core/actions/runs/3122291934/jobs/5064137152#step:12:27) - error: Expression of type "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" cannot be assigned to declared type "CoreSchema"
    Type "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" cannot be assigned to type "CoreSchema"
      "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" is incompatible with "AnySchema"
      "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" is incompatible with "BoolSchema"
      "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" is incompatible with "BytesSchema"
      "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" is incompatible with "DictSchema"
      "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" is incompatible with "FloatSchema"
      "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" is incompatible with "FunctionSchema"
      "dict[str, str | list[str | dict[str, str | int] | dict[str, str | float]]]" is incompatible with "FunctionPlainSchema"
    ... (reportGeneralTypeIssues)
  /home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py:19:26 - error: Expression of type "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" cannot be assigned to declared type "CoreSchema"
    Type "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" cannot be assigned to type "CoreSchema"
      "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" is incompatible with "AnySchema"
      "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" is incompatible with "BoolSchema"
      "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" is incompatible with "BytesSchema"
      "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" is incompatible with "DictSchema"
      "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" is incompatible with "FloatSchema"
      "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" is incompatible with "FunctionSchema"
      "dict[str, str | dict[str, dict[str, str | dict[str, dict[str, dict[str, str]]]]]]" is incompatible with "FunctionPlainSchema"
    ... (reportGeneralTypeIssues)
  /home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py:44:26 - error: Expression of type "dict[str, str | dict[str, str] | int]" cannot be assigned to declared type "CoreSchema"
    Type "dict[str, str | dict[str, str] | int]" cannot be assigned to type "CoreSchema"
      "dict[str, str | dict[str, str]]" is incompatible with "AnySchema"
      "dict[str, str | dict[str, str]]" is incompatible with "BoolSchema"
      "dict[str, str | dict[str, str]]" is incompatible with "BytesSchema"
      "dict[str, str | dict[str, str]]" is incompatible with "DictSchema"
      "dict[str, str | dict[str, str]]" is incompatible with "FloatSchema"
      "dict[str, str | dict[str, str]]" is incompatible with "FunctionSchema"
      "dict[str, str | dict[str, str]]" is incompatible with "FunctionPlainSchema"
    ... (reportGeneralTypeIssues)
  /home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py:56:26 - error: Expression of type "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" cannot be assigned to declared type "CoreSchema"
    Type "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" cannot be assigned to type "CoreSchema"
      "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" is incompatible with "AnySchema"
      "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" is incompatible with "BoolSchema"
      "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" is incompatible with "BytesSchema"
      "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" is incompatible with "DictSchema"
      "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" is incompatible with "FloatSchema"
      "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" is incompatible with "FunctionSchema"
      "dict[str, str | Type[Foo] | dict[str, str | bool | dict[str, dict[str, dict[str, str]]]]]" is incompatible with "FunctionPlainSchema"
    ... (reportGeneralTypeIssues)
  /home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py:62:26 - error: Expression of type "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" cannot be assigned to declared type "CoreSchema"
    Type "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" cannot be assigned to type "CoreSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" is incompatible with "AnySchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" is incompatible with "BoolSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" is incompatible with "BytesSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" is incompatible with "DictSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" is incompatible with "FloatSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" is incompatible with "FunctionSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str] | str] | dict[str, dict[str, str] | list[list[str | int] | list[str]]]]]" is incompatible with "FunctionPlainSchema"
    ... (reportGeneralTypeIssues)
  /home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py:72:26 - error: Expression of type "dict[str, str | ((bar: str) -> None) | dict[str, str]]" cannot be assigned to declared type "CoreSchema"
    Type "dict[str, str | ((bar: str) -> None) | dict[str, str]]" cannot be assigned to type "CoreSchema"
      "dict[str, str | ((bar: str) -> None) | dict[str, str]]" is incompatible with "AnySchema"
      "dict[str, str | ((bar: str) -> None) | dict[str, str]]" is incompatible with "BoolSchema"
      "dict[str, str | ((bar: str) -> None) | dict[str, str]]" is incompatible with "BytesSchema"
      "dict[str, str | ((bar: str) -> None) | dict[str, str]]" is incompatible with "DictSchema"
      "dict[str, str | ((bar: str) -> None) | dict[str, str]]" is incompatible with "FloatSchema"
      "dict[str, str | ((bar: str) -> None) | dict[str, str]]" is incompatible with "FunctionSchema"
      "dict[str, str | ((bar: str) -> None) | dict[str, str]]" is incompatible with "FunctionPlainSchema"
    ... (reportGeneralTypeIssues)
  /home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py:76:26 - error: Expression of type "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" cannot be assigned to declared type "CoreSchema"
    Type "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" cannot be assigned to type "CoreSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" is incompatible with "AnySchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" is incompatible with "BoolSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" is incompatible with "BytesSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" is incompatible with "DictSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" is incompatible with "FloatSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" is incompatible with "FunctionSchema"
      "dict[str, str | dict[str, dict[str, dict[str, str]] | dict[str, dict[str, str | dict[str, str | list[dict[str, str]]] | None]]]]" is incompatible with "FunctionPlainSchema"
    ... (reportGeneralTypeIssues)
  /home/runner/work/pydantic-core/pydantic-core/tests/test_typing.py:126:48 - error: Expression of type "dict[str, str]" cannot be assigned to declared type "CoreSchemaCombined"
    Type "dict[str, str]" cannot be assigned to type "CoreSchemaCombined"
      Type "dict[str, str]" cannot be assigned to type "CoreSchema"
        "dict[str, str]" is incompatible with "AnySchema"
        "dict[str, str]" is incompatible with "BoolSchema"
        "dict[str, str]" is incompatible with "BytesSchema"
        "dict[str, str]" is incompatible with "DictSchema"
        "dict[str, str]" is incompatible with "FloatSchema"
        "dict[str, str]" is incompatible with "FunctionSchema"
    ... (reportGeneralTypeIssues)
13 errors, 1 warning, 0 informations 
Completed in 3.761sec
WARNING: there is a new pyright version available (v1.1.[27](https://github.com/pydantic/pydantic-core/actions/runs/3122291934/jobs/5064137152#step:12:28)2).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`
```

</details>

To reproduce:

Clone the repo, then

```bash
cd pydantic-core
git checkout 1e24e5c
python3 -m venv env310
source env310/bin/activate
pip install pyright==1.1.246
pyright
```
(https://github.com/pydantic/pydantic-core/commit/1e24e5c4242dae54f816f83b32b271c985fe72d6 is the comment directly before the fix/workaround https://github.com/pydantic/pydantic-core/commit/ce0dce661d4d4cf313a607079b732d63a40f553a)

**Comment by samuelcolvin** [comment]
Bit of context here, mypy couldn't be used to check the types in [`core_schema.py`](https://github.com/pydantic/pydantic-core/blob/ce0dce661d4d4cf313a607079b732d63a40f553a/pydantic_core/core_schema.py) before they're recursive.

We therefore use pyright to test that static type analysis of these types matches the checks done when creating pydantic-core validators in rust.

We don't run pyright on the library code but rather to on a [test file](https://github.com/pydantic/pydantic-core/blob/ce0dce661d4d4cf313a607079b732d63a40f553a/tests/test_typing.py), until now pyright analysis has matched pydantic-core logic and my intuition well.

**Comment by erictraut** [comment]
OK, thanks for the additional instruction. I'm able to repro the problem. I'll investigate further.

**Comment by urayoru113** [comment]
Same issue

Can Pyright detect this?

![img](https://i.imgur.com/yzlBx5W.png)

**Comment by erictraut** [comment]
@urayoru113, the screen shot that you've posted is unrelated to this bug report. If you have questions about type checking or the use of pyright, please post those questions in the discussion section. If you would like to report a bug, please open a new issue and include a minimal self-contained code sample in text form.

**Comment by urayoru113** [comment]
@erictraut 
Thanks. I got the answer from https://github.com/microsoft/pyright/discussions/1355

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by samuelcolvin** [comment]
Great, thanks so much.

**Comment by erictraut** [comment]
This is included in pyright 1.1.279, which I just published. It will also be included in a future version of pylance.
